### DISCUSSION TITLE
Configuration for VSCode with MOOSE

### DISCUSSION THREAD

**Discussion by bielsnohr** [idaholab/moose#18307] [Q&A General] [answered]
Does anyone have experience using VSCode to develop MOOSE apps? I have been using Vim with YouCompleteMe to provide autocompletion via a `compile_commands.json` file, which is in turn created by a program `compiledb`. Too many links in the chain, and it falls down too often. The big problem is that `compiledb` does something strange with the makefile and touches file timestamps such that anytime I compile, is does a complete compilation of the whole MOOSE framework... time consuming to say the least.

So, I have been looking at moving to VSCode in the hopes that it would handle this sort of thing better, but it is not obvious how to get it set up with GNU Make projects. There is an extension called "Makefile Tools" but it is in preview (i.e. some sort of post beta) and hasn't worked out of the box for me. Suggestions?

**Comment by ajacquey** [comment, marked as answer]
Hi,

Here are the files I used for my MOOSE-Apps:

- `.vscode/task.json`: to build your app based on the Makefile. You can setup a keybinding for building the default build task (make test), I think by default it's ctrl + shift + b.
```
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            "label": "Makefike",
            "type": "shell",
            "command": "make",
            // start the build without prompting for task selection, use "group": "build" otherwise
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "options": {"shell": {"args": ["-ic"]}},
            "presentation": {
                "echo": true,
                "reveal": "always",
                "focus": false,
                "panel": "shared"
            },
            // arg passing example: in this case is executed with 4 processors
            "args": ["test", "-j4"],
            // Use the standard less compilation problem matcher.
            "problemMatcher": {
                "owner": "cpp",
                "fileLocation": ["absolute"],
                "pattern": {
                    "regexp": "^(.*):(\\d+):(\\d+):\\s+(warning|error):\\s+(.*)$",
                    "file": 1,
                    "line": 2,
                    "column": 3,
                    "severity": 4,
                    "message": 5
                }
            }
        },
        {
            "label": "Makefike debug",
            "type": "shell",
            "command": "make",
            // start the build without prompting for task selection, use "group": "build" otherwise
            "group": "none",
            "options": {"shell": {"args": ["-ic"]}},
            "presentation": {
                "echo": true,
                "reveal": "always",
                "focus": false,
                "panel": "shared"
            },
            // arg passing example: in this case is executed in DEBUG mode with 4 processors
            "args": ["METHOD=dbg", "-j4"],
            // Use the standard less compilation problem matcher.
            "problemMatcher": {
                "owner": "cpp",
                "fileLocation": ["absolute"],
                "pattern": {
                    "regexp": "^(.*):(\\d+):(\\d+):\\s+(warning|error):\\s+(.*)$",
                    "file": 1,
                    "line": 2,
                    "column": 3,
                    "severity": 4,
                    "message": 5
                }
            }
        },
        {
            "label": "Makefike clean",
            "type": "shell",
            "command": "make",
            // start the build without prompting for task selection, use "group": "build" otherwise
            "group": "none",
            "options": {"shell": {"args": ["-ic"]}},
            "presentation": {
                "echo": true,
                "reveal": "always",
                "focus": false,
                "panel": "shared"
            },
            // arg passing example: in this case is executed for clean
            "args": ["clean"],
            // Use the standard less compilation problem matcher.
            "problemMatcher": {
                "owner": "cpp",
                "fileLocation": ["absolute"],
                "pattern": {
                    "regexp": "^(.*):(\\d+):(\\d+):\\s+(warning|error):\\s+(.*)$",
                    "file": 1,
                    "line": 2,
                    "column": 3,
                    "severity": 4,
                    "message": 5
                }
            }
        },
        {
            "label": "Test",
            "type": "shell",
            "command": "./run_tests",
            // start the build without prompting for task selection, use "group": "build" otherwise
            "group": {
                "kind": "test",
                "isDefault": true
            },
            "options": {"shell": {"args": ["-ic"]}},
            "presentation": {
                "echo": true,
                "reveal": "always",
                "focus": false,
                "panel": "shared"
            },
            // arg passing example: in this case is executed with 4 processors
            "args": ["-j4"]
        }
    ]
}
```
- `.vscode/launch.json`: to configure the debugger
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python: Current File",
            "type": "python",
            "request": "launch",
            "program": "${file}",
            "console": "integratedTerminal"
        },
        {
            "name": "(lldb) Launch",
            "type": "cppdbg",
            "request": "launch",
            "program": "${workspaceFolder}/yourapp-dbg",
            "args": [
                "-i",
                "test/tests/simple_diffusion/simple_diffusion.i"
            ],
            "stopAtEntry": false,
            "cwd": "${workspaceFolder}",
            "environment": [],
            "externalConsole": true,
            "MIMode": "lldb",
            "logging": {
                "trace": true,
                "traceResponse": true,
                "engineLogging": true
            }
        }
    ]
}
```
- `.vscode/c_cpp_properties.json` for being able to navigate the C++ code. This is a working file for Linux. It assumes your app and moose folders are in the same directory and that you installed MOOSE using miniconda. You might have to change some paths depending on your configuration.
```
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}",
                "${workspaceFolder}/build/header_symlinks",
                "${workspaceFolder}/../moose/framework/build/header_symlinks",
                "${workspaceFolder}/../moose/modules/module_loader/build/header_symlinks",
                "${workspaceFolder}/../moose/framework/contrib/boost/include",
                "${workspaceFolder}/../moose/framework/contrib/hit",
                "${HOME}/miniconda3/envs/moose/include",
                "${HOME}/miniconda3/envs/moose/libmesh/include",
                "${workspaceFolder}/../moose/framework/contrib/json/include"
            ],
            "defines": [],
            // "macFrameworkPath": [
            //     "/System/Library/Frameworks",
            //     "/Library/Frameworks"
            // ],
            "compilerPath": "${HOME}/miniconda3/envs/moose/bin/mpicxx",
            "cStandard": "c11",
            "cppStandard": "c++17",
            "intelliSenseMode": "clang-x64",
            "compileCommands": "${workspaceFolder}/compile_commands.json",
            "browse": {
                "path": [
                    "${workspaceFolder}",
                    "${workspaceFolder}/build/header_symlinks",
                    "${workspaceFolder}/../moose/framework/build/header_symlinks",
                    "${workspaceFolder}/../moose/modules/module_loader/build/header_symlinks",
                    "${workspaceFolder}/../moose/framework/contrib/boost/include",
                    "${workspaceFolder}/../moose/framework/contrib/hit",
                    "${HOME}/miniconda3/envs/moose/include",
                    "${HOME}/miniconda3/envs/moose/libmesh/include"
                ]
            }
        }
    ],
    "version": 4
}
```
- For autocompletion of the input files, there is this [VSCode extension by Chris Sewell](https://marketplace.visualstudio.com/items?itemName=chrisjsewell.moose). It relies on the syntax files (.json and .yaml) you can generate from your executable. I do not have much experience with this extension though but may be someone else can comment on that.

Hope this helps.
Cheers,
Antoine

**Reply by bielsnohr** [reply]
Thanks for this! Really helpful. A minor addendum for those that might use this in the future: my IntelliSense didn't work until I specified that all `.h` files should be associated with C++ rather than C. Stick this in your directory `settings.json` if you encounter the same problem:

```json
{
    "files.associations": {
        "*.C": "cpp",
        "*.h": "cpp"
    }
}
```

**Comment by salaudeen-ya** [comment]
Has anyone used this configuration recently? The files I got from a colleague stopped working after certain modifications and upgrade in MOOSE.

**Reply by GiudGiud** [reply]
use the one on the moose website
https://mooseframework.inl.gov/help/development/VSCode.html

**Reply by salaudeen-ya** [reply]
Thank you!

I actually need something that can allow me to debug the input files (by stepping into the MOOSE codes) to know the order of execution. I would like to see the order of execution of the Shape functions and the Jacobians, and where they are computed, using the debug mode from the input file.

**Reply by GiudGiud** [reply]
see this page and use show_execution_order
https://mooseframework.inl.gov/syntax/Debug/index.html

**Reply by salaudeen-ya** [reply]
```
[Debug]
  show_execution_order = 'FORWARD'
  show_actions = true
[]
```
`show_execution_order = 'FORWARD'` gives `/testing.i:81: unused parameter 'Debug/show_execution_order'` error

**Reply by GiudGiud** [reply]
use ALWAYS
how old is your MOOSE?

**Reply by salaudeen-ya** [reply]
I used 'ALWAYS', same error. Pretty much all the options give the same error.

My MOOSE is very very recent.
```
Framework Information:
MOOSE Version:           git commit 430275f180 on 2023-06-19
LibMesh Version:         
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
Current Time:            Fri Jun 23 13:43:56 2023
Executable Timestamp:    Tue Jun 20 18:55:39 2023
```

**Reply by GiudGiud** [reply]
parameter must be misspelled then

**Reply by salaudeen-ya** [reply]
My Debug block in the input file is as given above, I don't think there's a spelling error in the `show_execution_order` based on what's on this page: https://mooseframework.inl.gov/syntax/Debug/index.html#debug-params

**Reply by andrey1887** [reply]
Hello. I try the similar c_cpp_properties.json. But i can't find compile_commands.json in my workspace folder.
How do i deal with this?
![изображение](https://github.com/idaholab/moose/assets/114737834/4fec8153-beb5-41b0-b9c6-c3fac68a26ab)

**Reply by GiudGiud** [reply]
I think you'll need to create that file yourself

**Reply by andrey1887** [reply]
Thank you. I currently create two files and add directives to it:
/home/and/projects/thin_walled/include/auxkernels/LocalAxisVectorAux.h
```
#pragma once
#include "Kernel.h"
```
/home/and/projects/thin_walled/src/auxkernels/LocalAxisVectorAux.C
`#include "VectorAuxKernel.h"`

What compile_comands.json might look like?

**Reply by andrey1887** [reply]
Or i need to use a snippet from [https://mooseframework.inl.gov/help/development/VSCode.html]() instead of suggested configuration?
(in c_cpp_properties.json)

**Reply by GiudGiud** [reply]
I just use that snippet for the user settings. 

I haven't set up the compile one (seemingly compile_commands.json). @dschwen did you set up that one

**Reply by ajacquey** [reply]
You can simply generate the file `compile_commands.json` with the `Makefile` of your MOSSE-based application:

```
make compile_commands.json
```

**Reply by andrey1887** [reply]
Thank you, @ajacquey . Now i have compile_commands.json. 
But my file src/auxkernels/LocalAxisVectorAux.C not found in it.
![image](https://github.com/idaholab/moose/assets/114737834/0c26524e-cac9-4726-93c9-96e6218b3d26)

**Reply by andrey1887** [reply]
@ajacquey, can you help me with this issue:
![изображение](https://github.com/idaholab/moose/assets/114737834/ff29f21a-c52d-47dd-9b4c-5e874ae3a985)
The same thing in .C and .h files. Thank you
